Source/WebDriver/Session.cpp

```diff
@@ -3144,4 +3146,139 @@ void Session::takeScreenshot(std::optional<String> elementID, std::optional<bool
+    // or moving the actual event processing to the browser, so we just relay it back to the client.
+    // https://bugs.webkit.org/show_bug.cgi?id=282976
+    // TODO Implement the full serialization algorithm
+    // https://bugs.webkit.org/show_bug.cgi?id=282977
+    auto messageText = params->getString("text"_s);
+    auto args = JSON::Array::create();
+    auto arg = JSON::Object::create();
+    arg->setString("type"_s, "string"_s);
+    arg->setString("value"_s, messageText);
+    args->pushObject(WTFMove(arg));
+
+    // TODO Get the source from the current realm record
+    // https://bugs.webkit.org/show_bug.cgi?id=282978
+
+    // TODO Get the current stacktrace for assert, error, trace, and warn messages
+    // https://bugs.webkit.org/show_bug.cgi?id=282979
+
+    auto entry = JSON::Object::create();
+    entry->setString("type"_s, "console"_s);
+    entry->setString("level"_s, level);
+    entry->setString("text"_s, messageText);
+    entry->setValue("timestamp"_s, *timestampValue);
+    entry->setString("method"_s, method);
+    entry->setArray("args"_s, args);
+
+    auto body = JSON::Object::create();
+    body->setObject("params"_s, WTFMove(entry));
+
+    if (eventIsEnabled("logEntryAdded"_s, { m_toplevelBrowsingContext.value() }))
+        emitEvent("log.entryAdded"_s, WTFMove(body));
+    // TODO Implement event buffering, to save the log entries for later emission when the user subscribes to it
+    // https://bugs.webkit.org/show_bug.cgi?id=282980
+}
+
+void Session::emitEvent(const String& eventName, RefPtr<JSON::Object>&& body)
+{
+    // https://w3c.github.io/webdriver-bidi/#emit-an-event
+    body->setString("type"_s, "event"_s);
+    body->setString("method"_s, eventName);
+    m_bidiServer->sendMessage(this->id(), body->toJSONString());
+}
+
+bool Session::eventIsEnabled(const String& eventName, const Vector<String>&)
+{
+    // https://w3c.github.io/webdriver-bidi/#event-is-enabled
+    HashSet<String> topLevelBrowsingContexts;
+    // FIXME Add support to subscribe to specific browsing contexts
+    // https://bugs.webkit.org/show_bug.cgi?id=282981
+
+    return m_globalEventSet.contains(eventName);
+}
+
+void Session::enableGlobalEvent(const String& eventName)
+{
+    m_globalEventSet.add(toInternalEventName(eventName));
+}
+
+void Session::disableGlobalEvent(const String& eventName)
+{
+    m_globalEventSet.remove(toInternalEventName(eventName));
+}
+
+String Session::toInternalEventName(const String& eventName)
+{
+    // The messages exchanged with the Browser (see Automation.json) can't have
+    // periods in the message name.
+    StringBuilder builder;
+    bool capitalizeNext = false;
+    for (unsigned i = 0; i < eventName.length(); i++) {
+        if (eventName[i] == '.') {
+            capitalizeNext = true;
+            continue;
+        }
+
+        if (capitalizeNext) {
+            builder.append(static_cast<char>(u_toupper(eventName[i])));
```

Comment:
all event names should be ASCII right?  if so, i think we could use `toASCIIUpper` instead

Reply:
The spec does not explicitly enforce this, but the currently defined (and supported) events are ascii, indeed.